src/tests/ftest/tags.py

```diff
@@ -0,0 +1,520 @@
+        if _list_len == 0:
+            return None
+        for _test in _list:
+            print(f'    {_test}')
+        if _list_len > 3:
+            remaining = _list_len - 3
+            _list = _list[:3] + [f"... (+{remaining})"]
+        _list_str = ", ".join(_list)
+        if not required:
+            # Print but do not fail
+            return None
+        return LintFailure(f"{_list_len} {message}: {_list_str}")
+
+    # Lint fails if any of the lists contain entries
+    errors = list(filter(None, [
+        _error_handler(non_unique_classes, 'non-unique test classes'),
+        _error_handler(non_unique_methods, 'non-unique test methods'),
+        _error_handler(test_wo_tags, 'tests without tags'),
+        _error_handler(tests_wo_class_as_tag, 'tests without class as tag', required=False),
+        _error_handler(tests_wo_method_as_tag, 'tests without method name as tag'),
+        _error_handler(tests_wo_hw_vm_manual, 'tests without HW, VM, or manual tag'),
+        _error_handler(tests_w_empty_tag, 'tests with an empty tag'),
+        _error_handler(tests_wo_a_feature_tag, 'tests without a feature tag')]))
+    if errors:
+        raise errors[0]
+
+
+def run_dump(paths=None):
+    """Dump the tags per test.
+
+    Formatted as
+        <file path>:
+          <class name>:
+            <method name> - <tags>
+
+    Args:
+        paths (list, optional): path(s) to get tags for. Defaults to all ftest python files
+    """
+    if not paths:
+        paths = all_python_files(FTEST_DIR)
+    for file_path, classes in iter(FtestTagMap(paths)):
+        short_file_path = re.findall(r'ftest/(.*$)', file_path)[0]
+        print(f'{short_file_path}:')
+        for class_name, functions in classes.items():
+            print(f'  {class_name}:')
+            all_methods = []
+            longest_method_name = 0
+            for method_name, tags in functions.items():
+                longest_method_name = max(longest_method_name, len(method_name))
+                all_methods.append((method_name, tags))
+            for method_name, tags in all_methods:
+                method_name_fm = method_name.ljust(longest_method_name, " ")
+                tags_fm = ",".join(sorted_tags(tags))
+                print(f'    {method_name_fm} - {tags_fm}')
+
+
+def files_to_tags(paths):
+    """Get the unique tags for paths.
+
+    Args:
+        paths (list): paths to get tags for
+
+    Returns:
+        set: set of test tags representing paths
+    """
+    # Get tags for ftest paths
+    ftest_tag_map = FtestTagMap(all_python_files(FTEST_DIR))
+    ftest_tag_set = ftest_tag_map.minimal_tags(paths)
+
+    # Future work will also get recommended tags for non-ftest files
+    return ftest_tag_set
+
+
+def run_list(paths):
+    """List unique tags for paths.
+
+    Args:
+        paths (list): paths to list tags of
+    """
+    tags = files_to_tags(paths)
+    print(' '.join(sorted(tags)))
+
+
+def run_unit(verbose):
```

Comment:
Well done for adding this, I in no way want to discourage writing tests.

It would be good if this could use pytest so that is obvious/common how to invoke it, and IMO it would be better in utest, not because it I don't want it to run in GHA but because I think we should be starting to migrate utest to GHA soon and a simple test like this would be an easy one to manage.  There's a danger in adding tests in *another* location that we loose track of where all the tests are.

Reply:
Good point. I haven't used pytest before, and I'm not very familiar with how to integrate into our utests. I'll look into that

Reply:
Basically you call `python -m pytest <file>` and it runs all method/classes in a file starting with `test_`.  I'm also not an expert but it's clearly the thing to use here.

Reply:
Well that seems stupidly simple. Let me try...

Reply:
Okay, it IS stupidly simple. Thanks :)
We don't seem to use pytest anywhere, and it's 3rd party so I didn't want to get blocked trying to get that installed.
So I modified the unit test stage to run it directly.

FWIW, there are a few other `.py` calls in the unit tests so probably a good idea to migrate those to pytest at some point

Reply:
We actually do have some pytest in the vos code but actually pytest isn't installed and therefore it isn't run but I do think it's the right direction to be headed.